### PR TITLE
feat: configurable strategy for background threads

### LINKED ISSUE #606: Consider using a CompletionQueue for background operations.
We have a few background operations to run:

- Polling for long running operations in `CreateInstance`, `CreateDatabase`, and the `Update*` versions.
- Keeping the sessions alive.

Maybe we should consider just having one thread (pool) and scheduling asynchronous requests using a `google::grpc_utils::CompletionQueue`.

### PR DESCRIPTION
**Pull request by coryan**
This change introduces an interface to represent the background
thread(s) used by a connection. I included two implementations in this
PR, one that automatically creates a single background thread, and
another that uses one or more threads provided by the application.

The ConnectionOptions class provides a factory for `BackgroundThreads`,
for those (we expect few) applications that will want to control the
background threads.

This fixes #606

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp-spanner/955)

### CODE REVIEW

File: google/cloud/spanner/connection_options.h [hunk 1]
```diff
@@ -152,13 +155,31 @@ class ConnectionOptions {
...
+   * Normally they just create a background thread and a `CompletionQueue` for
+   * this work, but the application may need more fine-grained control of their
+   * threads. In this case the application can provide the
+   */
+  ConnectionOptions& DisableBackgroundThreads(
+      google::cloud::grpc_utils::CompletionQueue const& cq);
+
+  using BackgroundThreadsFactory =
+      std::function<std::unique_ptr<BackgroundThreads>()>;
+  BackgroundThreadsFactory background_threads_factory() {
```
Comment (devjgm) [inline google/cloud/spanner/connection_options.h]: I'm not sure I understand how to use this method. It looks like every call to this getter will return a factory that, when invoked, will create a new "BackgroundThread", and by default, will use a new CompletionQueue. Is that right?

Would it be good to just expose access to a single CompletionQueue instance that everyone can use? But honestly, maybe I'm just not understanding how this works.

File: google/cloud/spanner/connection_options.h [hunk 2]
```diff
@@ -152,13 +155,31 @@ class ConnectionOptions {
    */
   grpc::ChannelArguments CreateChannelArguments() const;
 
+  /**
+   * Configure the connection to use @p cq for all background work.
+   *
+   * Connections need to perform background work on behalf of the application.
+   * Normally they just create a background thread and a `CompletionQueue` for
+   * this work, but the application may need more fine-grained control of their
+   * threads. In this case the application can provide the
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/spanner/connection_options.h]: Fixed.

File: google/cloud/spanner/background_threads.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ */
+class BackgroundThreads {
+ public:
+  virtual ~BackgroundThreads() = default;
+
+  /// The completion queue used for the background operations.
+  virtual grpc_utils::CompletionQueue cq() const = 0;
+
+  /// Terminate any automatically created threads.
+  virtual void Shutdown() = 0;
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/spanner/background_threads.h]: I am not sure. I *think* we will want to shut it down explicitly. But let's be conservative, I have removed the member function, we can always add it later if we really need it.

File: google/cloud/spanner/connection_options.h (hunk 1 above)
Comment (coryan) [reply google/cloud/spanner/connection_options.h]: > I'm not sure I understand how to use this method. It looks like every call to this getter will return a factory that, when invoked, will create a new "BackgroundThread", and by default, will use a new CompletionQueue. Is that right?

Yes.

> Would it be good to just expose access to a single CompletionQueue instance that everyone can use? But honestly, maybe I'm just not understanding how this works.

I think we decided to tie the lifetime of the automatically created threads to the lifetime of the `spanner::*Connection` object. That means the `*Connection` needs to know if it creates a thread and completion queue, and if it does create them it needs to shutdown the completion queue and join the thread in its destructor.  All those ifs suggested that this was a strategy given to the `*Connection` object.

In other words, we need to return something more than just the `CompletionQueue`, we need either (a) a thing that creates (when needed) the background thread for you, or (b) a `bool` that tells you if you should create a thread.

I went with option (a) because it seems wasteful to have every `spanner::*Connection` implement the "If I was asked to create a thread do this, otherwise do that".

File: google/cloud/spanner/connection_options_test.cc [hunk 4]
```diff
@@ -146,6 +146,47 @@ TEST(ConnectionOptionsTest, CreateChannelArguments_WithChannelPool) {
               StartsWith(options.user_agent_prefix()));
 }
 
+TEST(ConnectionOptionsTest, DefaultBackgroundThreads) {
+  auto options = ConnectionOptions(grpc::InsecureChannelCredentials());
+  auto background = options.background_threads_factory()();
+
+  using ms = std::chrono::milliseconds;
+
+  // Verify the background thread is not the main thread.
```
Comment (devbww) [inline google/cloud/spanner/connection_options_test.cc]: s/the main/this/?

File: google/cloud/spanner/internal/background_threads_impl.cc [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    : runner_([](grpc_utils::CompletionQueue cq) { cq.Run(); }, cq_) {}
+
+AutomaticallyCreatedBackgroundThreads::
+    ~AutomaticallyCreatedBackgroundThreads() {
+  Shutdown();
+}
+
+void AutomaticallyCreatedBackgroundThreads::Shutdown() {
+  cq_.Shutdown();
+  if (runner_.joinable()) runner_.join();
```
Comment (devbww) [inline google/cloud/spanner/internal/background_threads_impl.cc]: How might `runner_` have become detached if we don't provide any access to it?

File: google/cloud/spanner/internal/background_threads_test.cc [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+TEST(AutomaticallyCreatedBackgroundThreads, IsActive) {
+  AutomaticallyCreatedBackgroundThreads actual;
+
+  using ms = std::chrono::milliseconds;
+
+  auto expired = actual.cq().MakeRelativeTimer(ms(0));
+  EXPECT_EQ(std::future_status::ready, expired.wait_for(ms(100)));
+}
+
+/// @test Verify the background thread is not the main thread.
```
Comment (devbww) [inline google/cloud/spanner/internal/background_threads_test.cc]: s/the main/this/

File: google/cloud/spanner/internal/background_threads_impl.cc (hunk 5 above)
Comment (coryan) [reply google/cloud/spanner/internal/background_threads_impl.cc]: You can call `Shutdown()` twice, or `Shutdown()` and then the destructor.  In both cases the second call would throw:

https://en.cppreference.com/w/cpp/thread/thread/join

Comment (coryan) [reply google/cloud/spanner/internal/background_threads_impl.cc]: Calling `join()` (via `Shutdown()` for example) leaves `runner_` in a non-joinable state, in which case `std::thread::join` throws.

File: google/cloud/spanner/internal/background_threads_test.cc (hunk 6 above)
Comment (coryan) [reply google/cloud/spanner/internal/background_threads_test.cc]: Fixed.

File: google/cloud/spanner/connection_options_test.cc (hunk 4 above)
Comment (coryan) [reply google/cloud/spanner/connection_options_test.cc]: Fixed.

**Comment by coryan** [comment]
PTAL

File: google/cloud/spanner/internal/background_threads_impl.cc (hunk 5 above)
Comment (devbww) [reply google/cloud/spanner/internal/background_threads_impl.cc]: OK.  (I wonder whether this class should do something to ensure it only ever tries to shutdown the CompletionQueue and join the thread once, but I'm happy if this has the same effect.)
